packages/theme/middleware/checkout.js

```diff
@@ -0,0 +1,35 @@
+import { canEnterShipping, canEnterBilling, canEnterPayment, canEnterThankYou } from '../helpers';
+
+export default async ({ app, $vsf }) => {
+  const currentPath = app.context.route.fullPath.split('/checkout/')[1];
+
+  if (!currentPath) return;
+
+  const cart = await $vsf.$vendure.api.getCart();
+  const activeCart = cart?.data?.activeOrder;
+
+  if (!cart?.data || !activeCart) return;
+
+  switch (currentPath) {
+    case 'shipping':
+      if (!canEnterShipping(activeCart)) {
+        app.context.redirect('/');
+      }
+      break;
+    case 'billing':
+      if (!canEnterBilling(activeCart)) {
+        app.context.redirect('/');
+      }
+      break;
+    case 'payment':
+      if (!canEnterPayment(activeCart)) {
+        app.context.redirect('/');
+      }
+      break;
+    case 'thank-you':
+      if (!canEnterThankYou(app.context)) {
+        app.context.redirect('/');
+      }
+      break;
+  }
```

Comment:
Maybe better option is to have just if.
```suggestion
    if(currentPath === 'shipping' && !canEnterShipping(activeCart))  {
      app.context.redirect('/');
      return;
    }
     if(currentPath === 'billing' && !canEnterBilling(activeCart))  {
      app.context.redirect('/');
      return;
    }
     if(currentPath === 'payment' && !canEnterPayment(activeCart))  {
      app.context.redirect('/');
      return;
    }
     if(currentPath === 'thank-you' && !canEnterThankYou(app.context))  {
      app.context.redirect('/');
      return;
    }
  }
```

Also IMO good idea is to replace strings via const or enum values

Reply:
const or enum values sound cool for me too